app/syncables.go

```diff
@@ -101,22 +172,65 @@ func (a *App) DeleteGroupConstrainedMemberships() error {
 		)
 	}
 
-	teamMembers, appErr := a.TeamMembersToRemove()
-	if appErr != nil {
-		return appErr
+	return nil
+}
+
+// SyncSyncableRoles updates the SchemeAdmin field value of the given syncable's members based on the configuration of
+// the member's group memberships and the configuration of those groups to the syncable. This method should only
+// be invoked on group-synced (aka group-constrained) syncables.
+func (a *App) SyncSyncableRoles(syncableID string, syncableType model.GroupSyncableType) *model.AppError {
+	permittedAdmins, err := a.Srv.Store.Group().PermittedSyncableAdmins(syncableID, syncableType)
+	if err != nil {
+		return err
 	}
 
-	for _, userTeam := range teamMembers {
-		err := a.RemoveUserFromTeam(userTeam.TeamId, userTeam.UserId, "")
-		if err != nil {
-			return err
-		}
+	a.Log.Info(
+		fmt.Sprintf("Permitted admins for %s", syncableType),
+		mlog.String(strings.ToLower(fmt.Sprintf("%s_id", syncableType)), syncableID),
+		mlog.Any("permitted_admins", permittedAdmins),
+	)
+
+	var updateFunc func(string, []string, store.Equality, bool) *model.AppError
+
+	switch syncableType {
+	case model.GroupSyncableTypeTeam:
+		updateFunc = a.Srv.Store.Team().UpdateMembersRole
+	case model.GroupSyncableTypeChannel:
+		updateFunc = a.Srv.Store.Channel().UpdateMembersRole
+	default:
+		return model.NewAppError("App.SyncSyncableRoles", "groups.unsupported_syncable_type", map[string]interface{}{"Value": syncableType}, "", http.StatusInternalServerError)
+	}
 
-		a.Log.Info("removed teammember",
-			mlog.String("user_id", userTeam.UserId),
-			mlog.String("team_id", userTeam.TeamId),
-		)
+	err = updateFunc(syncableID, permittedAdmins, store.Equals, true)
+	if err != nil {
+		return err
+	}
+
+	err = updateFunc(syncableID, permittedAdmins, store.NotEquals, false)
```

Comment:
I'm wondering if "UpdateMembersRole" should run both queries, rather than calling twice.  Is there ever a time you would only set one set and not the other?

Reply:
My original thinking was to keep the data store "dumber" but now that you've question it I'm changing my mind... it could be a single SQL `UPDATE` where everything `IN` has a boolean value and everything `NOT IN` has the opposite value. I'll make this change. Thanks.

Reply:
I have [changed the `UpdateMembersRole`](https://github.com/mattermost/mattermost-server/pull/13361/commits/8670e7815d6dc13639b08350f603fec9c1218ab1) methods in a separate PR that this is branched off of. It's using a SQL case statement to conditionally update the `SchemeAdmin` column value in a single query.